### ISSUE TITLE
Unit/Functional Testing documentation

### ISSUE THREAD

**Issue by aligentjim** [orocrm/documentation#236]
I'm really struggling with the test related documentation (https://www.orocrm.com/documentation/1.10/book/functional-tests).  We're using OroCRM 1.10, hence I'm referring to that documentation, but the 2.0 docs for this topic look basically the same, so I'm assuming the problems I'm having haven't been addressed there.

Behat is installed via composer, but the docs do not mention Behat, instead focussing exclusively on phpUnit.  Which testing tool is actually used, Behat or phpUnit?

Assuming the docs are correct, and I should be using phpUnit, how do I install it?  The linked Symfony documentation says that installation from PHAR is preferred, but then when I try to run the PHAR installed phpUnit (5.7 - The last one with PHP 5.x support), I get...

```
$ phpunit src/Spendless/
Class "JohnKary\PHPUnit\Listener\SpeedTrapListener" does not exist
```

This class is defined as a listener in the phpunit.xml.dist that ships with OroCRM, but obviously not bundled in the PHAR, not is it referenced in the composer.json that comes with OroCRM.

If I attempt to install phpUnit via composer, I get a dependency issue caused by the existing OroCRM dependencies...

```
$ ../utils/composer require --dev phpunit/phpunit 5.7.*
./composer.json has been updated
Loading composer repositories with package information
Updating dependencies (including require-dev)
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - Conclusion: don't install phpunit/phpunit 5.7.19
    - Conclusion: don't install phpunit/phpunit 5.7.18
    - Conclusion: don't install phpunit/phpunit 5.7.17
    - Conclusion: don't install phpunit/phpunit 5.7.16
    - Conclusion: don't install phpunit/phpunit 5.7.15
    - Conclusion: don't install phpunit/phpunit 5.7.14
    - Conclusion: don't install phpunit/phpunit 5.7.13
    - Conclusion: don't install phpunit/phpunit 5.7.12
    - Conclusion: don't install phpunit/phpunit 5.7.11
    - Conclusion: don't install phpunit/phpunit 5.7.10
    - Conclusion: don't install phpunit/phpunit 5.7.9
    - Conclusion: don't install phpunit/phpunit 5.7.8
    - Conclusion: don't install phpunit/phpunit 5.7.7
    - Conclusion: don't install phpunit/phpunit 5.7.6
    - Conclusion: don't install phpunit/phpunit 5.7.5
    - Conclusion: don't install phpunit/phpunit 5.7.4
    - Conclusion: don't install phpunit/phpunit 5.7.3
    - Conclusion: don't install phpunit/phpunit 5.7.2
    - Conclusion: don't install phpunit/phpunit 5.7.1
    - Conclusion: don't install phpunit/phpunit 5.7.0
    - Conclusion: remove sebastian/comparator 1.2.2
    - Installation request for sebastian/exporter (locked at 1.2.2) -> satisfiable by sebastian/exporter[1.2.2].
    - Installation request for phpunit/phpunit 5.7.* -> satisfiable by phpunit/phpunit[5.7.0, 5.7.1, 5.7.10, 5.7.11, 5.7.12, 5.7.13, 5.7.14, 5.7.15, 5.7.16, 5.7.17, 5.7.18, 5.7.19, 5.7.2, 5.7.3, 5.7.4, 5.7.5, 5.7.6, 5.7.7, 5.7.8, 5.7.9, 5.7.x-dev].
    - Conclusion: don't install sebastian/comparator 1.2.2
    - phpunit/phpunit 5.7.x-dev requires sebastian/comparator ^1.2.4 -> satisfiable by sebastian/comparator[1.2.4, 1.2.x-dev].
    - Can only install one of: sebastian/comparator[1.2.4, 1.2.2].
    - Can only install one of: sebastian/comparator[1.2.x-dev, 1.2.2].
    - Installation request for sebastian/comparator (locked at 1.2.2) -> satisfiable by sebastian/comparator[1.2.2].


Installation failed, reverting ./composer.json to its original content.
``` 

So long story short, how does one go about setting up a dev environment to run Unit/Functional tests.  If there's good documentation on the internet, can a link to that documentation be included in the OroCRM documentation.  If it doesn't exist, what's the "right" way to get set up?  If someone provides notes on a working configuration, I'm happy to turn that into "real" documentation and create a PR.

**Comment by mkudelya** [comment]
Hi

We are using PHPUnit for Unit/Functional testing. For example when i have ran `phpunit48.phar -c app/ vendor/oro/platform/src/Oro/Bundle/UserBundle/Tests/Unit/Provider` i got the following output:

    bash-3.2$ phpunit48.phar -c 
    app/vendor/oro/platform/src/Oro/Bundle/UserBundle/Tests/Unit/Provider

    PHPUnit 4.8.23 by Sebastian Bergmann and contributors.

    I.......

    Time: 134 ms, Memory: 10.00Mb

    OK, but incomplete, skipped, or risky tests!
    Tests: 8, Assertions: 18, Incomplete: 1.

Also please read [the behat documentation](https://github.com/orocrm/platform/blob/master/src/Oro/Bundle/TestFrameworkBundle/Resources/doc/reference/behat-tests.md).

**Comment by aligentjim** [comment]
Interesting, so when I use the Phar installed phpunit and run with the `-c app/` parameter I have a more successful experience.  I think I was being confused by the fact that there are `phpunit.xml.dist` files in both the root folder of the repo (https://github.com/orocrm/crm-application/blob/master/phpunit.xml.dist) and the `app/` folder (https://github.com/orocrm/crm-application/blob/master/app/phpunit.xml.dist).  Is there a need for both of these to exist.  Removing the "broken" one in the root would simplify things.

**Comment by mkudelya** [comment]
Hi @aligentjim 

We are using phpunit.xml.dist file in the root folder for vendor testing and /app/phpunit.xml.dist for testing inside the /src folder. Please, modify (add/delete a test suit) phpunit.xml for your goals.
Just for your attention, in 2.* version of orocrm phpunit doesn't produce the `Class "JohnKary\PHPUnit\Listener\SpeedTrapListener" does not exist` error.

**Comment by sergeyz** [comment]
@aligentjim please do `composer install` without `--no-dev` option and use `vendor/bin/phpunit` binary. Starting from 1.10 version we provide all tools we use for development and CI in `require-dev section` https://github.com/orocrm/platform-application/blob/1.10/composer.json#L21-L30

**Comment by aligentjim** [comment]
Since raising this ticket, we've upgraded the customer to 2.1 and we've been able to get unit tests up and running without issues there.  Since this relates to an older version, there's probably no value in spending more time on it.
